[not ready for merge] Userspace stack tracing from kernel programs

brenns10 opened the pull request:
This has always been a bit of a far-off idea, but with the module API it's working, some of the time (definitely not all of the time). So I thought I'd share it to see if some of the tweaks necessary to make it happen would be reasonable.

Basically, there's a GDB script called `pstack` that attaches GDB and takes a stack trace of a program. You can also use `/proc/$PID/stack` to get the kernel stack (assuming it's running in kernel mode or blocked). I was hoping to come up with a way to replicate that behavior in drgn, in a way that would work against `/proc/kcore` or `/proc/vmcore`. Essentially, it would allow you to get userspace stack traces from the crashed kernel (but not necessarily the whole core dump, like `contrib/gcore.py`) in the kdump kernel just before or after dumping the vmcore. (Presumably, userspace pages are filtered for 99.9% of all kdump configurations, so you'd need to run this while `/proc/kcore` is still available).

The main part of this requires creating a custom program that has a memory reader, as well as specifying all the required Modules, their biases, and their address ranges. From there, you can get the userspace `struct pt_regs` from the kernel program, copy it to the user program, and then unwind the stack.

To do this, I've needed to tweak drgn a bit:

1. Python memory readers may return FaultError, but the resulting `drgn_error` has the wrong error code. So I added a special-case to pass through fault-errors back to the drgn error. This could be made more general, but I don't actually think it would be good to do that generally.
2. I made loaded & debug file biases writable, so that I could update the biases.
3. I included a crude patch to support `.gnu_debugdata` which is helpful for my use case.
4. I needed to rip out the compatibility checks for stack tracing, because otherwise drgn would fail saying that stack tracing is not supported for this program.

At the end of the day, I'm quite confident that none of this is ready for merge, but I did wonder if any of the individual changes make sense to include? 

For fun, here's the result of running the `contrib/pstack.py` script against the current bash process:
```
$ python -m drgn -q --kernel-dir ~/vmlinux_repo/$(uname -r) -k contrib/pstack.py $$
#0  context_switch (kernel/sched/core.c:5328:2)
#1  __schedule (kernel/sched/core.c:6693:8)
#2  __schedule_loop (kernel/sched/core.c:6770:3)
#3  schedule (kernel/sched/core.c:6785:2)
#4  do_wait (kernel/exit.c:1697:3)
#5  kernel_wait4 (kernel/exit.c:1851:8)
#6  __do_sys_wait4 (kernel/exit.c:1879:13)
#7  do_syscall_x64 (arch/x86/entry/common.c:52:14)
#8  do_syscall_64 (arch/x86/entry/common.c:89:7)
#9  entry_SYSCALL_64+0xaf/0x14c (arch/x86/entry/entry_64.S:121)
#10 0x7ff5ba4d8b7a
------ userspace ---------
#0  wait4+0x1a/0xab
#1  waitchld.constprop.0+0xbb/0xa5f
#2  wait_for+0x4ca/0xc0e
#3  execute_command_internal+0x2768/0x2ef6
#4  execute_command+0xc8/0x1b8
#5  reader_loop+0x289/0x3d9
#6  main+0x15be/0x198b
#7  __libc_start_call_main+0x80/0xac
#8  __libc_start_main@@GLIBC_2.34+0x80/0x148
#9  _start+0x25/0x26
#10 ???
```

brenns10 commented:
Still a draft - the `.gnu_debugdata` support is still a big hack. Though, if desired, I can yank that out to a separate pull request.

I've dropped the hack to allow setting the file bias, and I've also dropped the change that allows loading non-debug, non-loadable files for "extra modules". Neither are necessary now.

I was able to make drgn get the correct file bias for the main module, but unfortunately I had to load it as an "extra module". Otherwise, drgn mis-computes the file bias (see comment in code). AFAICT, that doesn't impact functionality much.

I still had to tweak the stack trace code, but I didn't rip out/uncomment the checks, just loosened one.

I made a few fixes to the contrib script:
- It now checks the inode number of the vma and file. That way it can warn the user if a program or shared library has been updated and likely won't be usable. I had actually spent a lot of time trying to figure out why this script wasn't working on so many of my processes, and it turned out that this was the reason 🙃 
- I fixed the detection of the dynamic address for shared libraries.

brenns10 commented:
I've gotten `.gnu_debugdata` to the point where I'm happy with it. The biggest issue was refactoring the ELF symbol finder so it supports iterating through multiple symbol tables. I did this in a separate commit to make it easier to review the actual `.gnu_debugdata` changes in isolation from them. It still needs tests, so it's not really ready for review yet -- sorry for the noise.

brenns10 commented:
Ok, it's been a bit noisy but I now think this is in a good state. I've added testing for `.gnu_debugdata` by repeating the same ELF symbol tests, but with the symbols split between the loaded file and its `.gnu_debugdata`.

osandov commented on libdrgn/python/error.c:
```diff
@@ -88,6 +111,12 @@ struct drgn_error *drgn_error_from_python(void)
 	}
 
+	if ((PyTypeObject *)exc_type == &FaultError_type && exc_value) {
+		exc_message = NULL;
```
Modernizing this function would make this less awkward. Something like:

```diff
diff --git a/libdrgn/python/error.c b/libdrgn/python/error.c
index 1d36a2e9..572c507b 100644
--- a/libdrgn/python/error.c
+++ b/libdrgn/python/error.c
@@ -75,46 +75,38 @@ void clear_drgn_in_python(void)
 
 struct drgn_error *drgn_error_from_python(void)
 {
-	PyObject *exc_type, *exc_value, *exc_traceback, *exc_message;
-	const char *type, *message;
-	struct drgn_error *err;
-
+	_cleanup_pydecref_ PyObject *exc_type, *exc_value, *exc_traceback;
 	PyErr_Fetch(&exc_type, &exc_value, &exc_traceback);
 	if (!exc_type)
 		return NULL;
 
 	if (drgn_in_python) {
 		PyErr_Restore(exc_type, exc_value, exc_traceback);
+		exc_type = exc_value = exc_traceback = NULL;
 		return &drgn_error_python;
 	}
 
-	type = ((PyTypeObject *)exc_type)->tp_name;
+	const char *type = ((PyTypeObject *)exc_type)->tp_name;
+	_cleanup_pydecref_ PyObject *exc_message = NULL;
+	const char *message;
 	if (exc_value) {
 		exc_message = PyObject_Str(exc_value);
 		message = exc_message ? PyUnicode_AsUTF8(exc_message) : NULL;
 		if (!message) {
-			err = drgn_error_format(DRGN_ERROR_OTHER,
-						"%s: <exception str() failed>", type);
-			goto out;
+			PyErr_Clear();
+			return drgn_error_format(DRGN_ERROR_OTHER,
+						 "%s: <exception str() failed>", type);
 		}
 	} else {
-		exc_message = NULL;
 		message = "";
 	}
 
 	if (message[0]) {
-		err = drgn_error_format(DRGN_ERROR_OTHER, "%s: %s", type,
-					message);
+		return drgn_error_format(DRGN_ERROR_OTHER, "%s: %s", type,
+					 message);
 	} else {
-		err = drgn_error_create(DRGN_ERROR_OTHER, type);
+		return drgn_error_create(DRGN_ERROR_OTHER, type);
 	}
-
-out:
-	Py_XDECREF(exc_message);
-	Py_XDECREF(exc_traceback);
-	Py_XDECREF(exc_value);
-	Py_DECREF(exc_type);
-	return err;
 }
 
 void *set_drgn_error(struct drgn_error *err)
```
Bonus missing `PyErr_Clear()` bug fix in there! Feel free to split that into its own commit or fold it in at your preference.

brenns10 replied:
Yeah, I got lazy on this, I'll do the full fix, thanks!

osandov commented on libdrgn/stack_trace.c:
```diff
@@ -1212,10 +1214,11 @@ static struct drgn_error *drgn_get_stack_trace(struct drgn_program *prog,
-		   && !drgn_program_is_userspace_core(prog)) {
-		return drgn_error_create(DRGN_ERROR_NOT_IMPLEMENTED,
-					 "stack unwinding is not supported for this program");
+	} else if (!warned_unsupported &&
+		   !(prog->flags & DRGN_PROGRAM_IS_LINUX_KERNEL)
+		     && !drgn_program_is_userspace_core(prog)) {
+		drgn_log_warning(prog, "warning: stack unwinding is not supported for custom programs");
+		warned_unsupported = true;
```
The only reason we can't get stack traces of other program types is that we can't get their initial registers. We could still allow unwinding via `struct pt_regs`. Maybe we can move the original checks to `drgn_get_initial_registers()` in the !pt_regs case instead?

brenns10 replied:
This is a _much_ better approach, thank you!

osandov reviewed (changes requested):
I looked at "libdrgn: python: pass fault errors through" and "libdrgn: stack_trace: allow unwinding custom programs" to start with, because I'd love to merge those independently of the rest of this. I'll look at the rest later.

brenns10 commented on libdrgn/python/error.c:
```diff
@@ -73,6 +73,29 @@ void clear_drgn_in_python(void)
+	message = PyObject_GetAttrString(exc_value, "message");
+	if (!message) {
+		PyErr_Clear();
+		goto out;
```
Yeah, this was silly. I think I meant to handle this by catching the `NULL` return value in `drgn_error_from_python()` so that it could continue into the normal handling case, but I never added that part. I've done so here.

brenns10 commented:
Feel free to cherry-pick what you want if you're happy with those commits and I can always drop them in the rebase.

osandov reviewed (commented):
Thanks, I cherry-picked the error handling patches. I have one question about "libdrgn: stack_trace: allow unwinding anything with a pt_regs" before I cherry-pick that one.

brenns10 commented on libdrgn/stack_trace.c:
```diff
@@ -1209,14 +1212,6 @@ static struct drgn_error *drgn_get_stack_trace(struct drgn_program *prog,
-	if (drgn_program_is_userspace_process(prog)) {
-		return drgn_error_create(DRGN_ERROR_NOT_IMPLEMENTED,
-					 "stack unwinding is not yet supported for live processes");
-	} else if (!(prog->flags & DRGN_PROGRAM_IS_LINUX_KERNEL)
-		   && !drgn_program_is_userspace_core(prog)) {
-		return drgn_error_create(DRGN_ERROR_NOT_IMPLEMENTED,
-					 "stack unwinding is not supported for this program");
-	}
```
Yeah, I did mean to drop that check! By my reading of `drgn_get_initial_registers()`, any kernel cores would be well handled, and any object that resembles a pt_regs would get handled too. Then there's just the question of userspace processes, userspace core dumps, and maybe some mysterious other kind of program not implemented yet... There's an assertion that crashes drgn for live programs within `drgn_program_cache_core_dump_threads()`, which we need to guard against, but all live programs at this point in execution _should_ be userspace processes. In the other situations we would end up at the "thread not found".

And as far as I can see, the only way to get there from Python is to do a `Thread.stack_trace()` operation -- any `Program.stack_trace()` would have been handled above, either for integers by interpreting them as a thread ID, or for other type kinds by looking for task structs or pt_regs.

So if you're good with it, I did think it was safe and try to think through the cases :)

osandov replied:
Great, that's what I figured, just wanted to double check. Just merged that one, too!

brenns10 commented:
(No changes except to rebase on the latest main & drop the cherry-picks.)

brenns10 commented:
Rebased against the latest main branch, and also dropped the contrib script, since it doesn't actually depend on `.gnu_debugdata` -- it's just a nice-to-have.

osandov commented on libdrgn/elf_symtab.c:
```diff
@@ -46,6 +58,98 @@ static struct drgn_error *find_elf_file_symtab(struct drgn_elf_file *file,
+			return drgn_error_lzma(ret);
+
+		bytes_decoded = (size_t)stream.next_out - (size_t)data;
+		if (ret == LZMA_STREAM_END)
+			datasize = bytes_decoded;
+		else
+			datasize *= 2;
+
+		data = realloc(data, datasize);
+		if (!data)
+			return &drgn_enomem;
+
+		stream.next_out = (uint8_t *)data + bytes_decoded;
+		stream.avail_out = datasize - bytes_decoded;
+		if (ret == LZMA_STREAM_END)
+			break;
```
A few things here:

* If the `realloc()` fails, you're leaking `data`.
* If the `realloc()` fails in the `LZMA_STREAM_END` case, it doesn't need to be fatal.
* Should again be paranoid about `datasize` overflow.

Maybe something like:

```suggestion
		if (ret == LZMA_STREAM_END) {
			// Try shrinking to fit.
			void *tmp = realloc(data, bytes_decoded);
			if (tmp)
				data = tmp;
			break;
		} else if (__builtin_mul_overflow(datasize, 2U, &datasize)) {
			return &drgn_enomem;
		} else {
			void *tmp = realloc(data, datasize);
			if (!tmp)
				return &drgn_enomem;
			data = tmp;
			stream.next_out = (uint8_t *)data + bytes_decoded;
			stream.avail_out = datasize - bytes_decoded;
		}
```

osandov commented on libdrgn/elf_symtab.c:
```diff
@@ -46,6 +58,98 @@ static struct drgn_error *find_elf_file_symtab(struct drgn_elf_file *file,
+	stream.next_in = gnu_debugdata_data->d_buf;
+	stream.avail_in = gnu_debugdata_data->d_size;
+
+	size_t datasize = 2 * gnu_debugdata_data->d_size;
```
I assume this is just a heuristic for a good initial size to start with? Please add a comment to that effect.

Also, it's probably not possible for this to overflow in practice, but I'd rather be paranoid. Maybe something like:

```suggestion
	size_t datasize;
	if (__builtin_mul_overflow(gnu_debugdata_data->size, 2U, &datasize)
		datasize = gnu_debugdata_data->size;
```

Also, `capacity` might be a better name.

brenns10 replied:
Yeah, it's just a heuristic guess. And honestly, I'll happily get rid of the guess and just start with the same size of the input, and then all the overflow checks can be done in the same place within the loop.

osandov commented:
I just thought of a problem with the appending to an array of symbol tables approach: in the case that we first find a dynamic symbol table, then later find a full symbol table, the full symbol table is supposed to replace the dynamic one (although as I noted earlier, I think that check is buggy right now). With the array approach, that is currently going to append the full symbol table and keep both. So it might be better to explicitly have a main/normal symbol table and a gnu_debugdata symbol table instead of an array.

brenns10 commented:
Yeah, that makes a lot of sense and shouldn't be a difficult change. Thanks for taking a look at the details here, I definitely missed a few of the intricacies of the ELF symbol finder.

osandov commented:
My fixes are in the `symtab-fixes` branch waiting for tests to pass.

brenns10 commented:
Argh sorry, pushed to the wrong remote, please ignore..

brenns10 commented on libdrgn/debug_info.h:
```diff
@@ -240,7 +240,7 @@ struct drgn_module {
 	/** ORC unwinder information. */
 	struct drgn_module_orc_info orc;
 	/** ELF symbol table. */
```
I think the better approach here is not to use an array anyway. I'll add a new symbol table named `gnu_debugdata_symtab` which will be much clearer, and then the `drgn_module_elf_symbols_search()` will have the "unrolled loop" which will also probably be clearer to read as well.

brenns10 commented on libdrgn/elf_symtab.c:
```diff
@@ -371,43 +387,44 @@ drgn_module_elf_symbols_search(struct drgn_module *module, const char *name,
 
-			max_end_addr = max(max_end_addr,
-					   elf_sym_addr + elf_sym.st_size);
+			state->max_end_addr = max(state->max_end_addr,
+						  elf_sym_addr + elf_sym.st_size);
```
I'm not sure I agree. Can you explain the problem this solves?

Here's the issue I was thinking of when keeping `max_end_addr` across both symbol tables. Suppose we have the following symbols:
```
ADDR  SIZE  NAME      WHICH SYMTAB?
1000    50  symbol_a  .dynsym
1050     0  asm_func  .dynsym
1100    50  symbol_b  .gnu_debugdata
1160    40  symbol_c  .gnu_debugdata
1200   100  symbol_d  .dynsym
```

Now consider an address search for address 1155. This is between symbol_b and symbol_c, and not contained by either. Maybe symbol_c had some alignment requirement that left a gap between them.

If we don't share `max_end_addr` across both symbol tables, then `asm_func` would be returned as a valid result. It is sizeless, and valid among the symbols in .dynsym. Yet if we look at `symbol_b` (only present in the `.gnu_debugdata`), this would imply that `asm_func` only extends 50 bytes, and shouldn't be considered a valid sizeless symbol containing the address 1155.

osandov replied:
Yup, I think you're right, sorry about that!

brenns10 commented on libdrgn/elf_symtab.c:
```diff
@@ -46,6 +58,98 @@ static struct drgn_error *find_elf_file_symtab(struct drgn_elf_file *file,
+	_cleanup_free_ void *data = malloc(datasize);
+
+	stream.next_out = data;
+	stream.avail_out = gnu_debugdata_data->d_size;
```
Yeah, thanks for the catch!

brenns10 commented:
Ok, local tests are passing, pre-commit is passing, and I think I've resolved everything from your review -- thanks for all the comments, they really improved things. I also ended up cleaning up `find_module_elf_symtab()` a bit more, so that:

1. If a full symtab gets loaded after a `.gnu_debugdata`, then we clean up `gnu_debugdata_file`, freeing up its memory. This is mostly just an optimization, I doubt it would really get hit too frequently. But we have no idea what could be added into these `.gnu_debugdata` ELF files in the future, so I think it's good to avoid holding them in memory more than we need.
2. It should be impossible to overwrite an existing `gnu_debugdata_file` or symbol table. However, if we did overwrite them, we would leak the extracted file, so I added an assertion to that effect. Though I don't know if that's over-zealous :)

brenns10 commented:
(Sorry for the re-push. Cleaned up an old unused variable from commit 1 which I could have sworn I cleaned up in my rebase, but clearly I missed.)

brenns10 commented on libdrgn/elf_symtab.c:
```diff
@@ -45,6 +58,106 @@ static struct drgn_error *find_elf_file_symtab(struct drgn_elf_file *file,
+	if (!elf)
+		return drgn_error_libelf();
+
+	err = drgn_elf_file_create(module, module->loaded_file->path, -1, data, elf,
```
You know, I should have done this myself. I did have an issue while developing where an error was reported and the filename didn't indicate that it was from the `.gnu_debugdata` file, which did indeed confuse me. I think I failed to change the path because at that moment I wasn't feeling like doing some C string manipulation 🤣

brenns10 commented on libdrgn/elf_symtab.c:
```diff
@@ -148,6 +272,56 @@ find_module_elf_symtab(struct drgn_module *module)
+				if (err)
+					return err;
+
+				module->elf_symtab_pending_files = 0;
```
It's not actually equivalent to the full symbol table.  The `.gnu_debugdata` section only contains the function symbols necessary for stack tracing, since that's all that's mentioned in the goals for [MiniDebuginfo](https://fedoraproject.org/wiki/Features/MiniDebugInfo). I believe they wanted to minimize it as much as possible. To illustrate:
```
stepbren at stepbren-ol9 in ~/repos/drgn (module_tweaks)
$ eu-readelf -s /usr/lib/debug/bin/bash-5.1.8-9.el9.x86_64.debug  | tail -n 5
14672: 0000000000000000      0 FUNC    GLOBAL DEFAULT    UNDEF mktemp@GLIBC_2.2.5
14673: 0000000000083950    377 FUNC    GLOBAL DEFAULT       15 discard_unwind_frame
14674: 0000000000158444      4 OBJECT  GLOBAL DEFAULT       26 _rl_undo_group_level
14675: 0000000000158740      4 OBJECT  GLOBAL DEFAULT       26 _rl_mark_modified_lines
14676: 00000000000da0d0    129 FUNC    GLOBAL DEFAULT       15 rl_history_substr_search_backward

stepbren at stepbren-ol9 in ~/repos/drgn (module_tweaks)
$ eu-readelf -s /bin/bash  | tail -n 5
 2512: 0000000000068370     22 FUNC    GLOBAL DEFAULT       15 restore_pgrp_pipe
 2513: 0000000000146044      4 OBJECT  GLOBAL DEFAULT       25 interactive_comments
 2514: 00000000000fa6d0    651 FUNC    GLOBAL DEFAULT       15 tilde_expand_word
 2515: 0000000000087860    323 FUNC    GLOBAL DEFAULT       15 array_shift
 2516: 000000000014f094      4 OBJECT  GLOBAL DEFAULT       26 history_lines_this_session

stepbren at stepbren-ol9 in ~/repos/drgn (module_tweaks)
$ eu-readelf --elf-section=.gnu_debugdata -s /bin/bash  | tail -n 5
 1868: 0000000000089690    575 FUNC    GLOBAL DEFAULT       15 add_alias
 1869: 00000000000b1990   1484 FUNC    GLOBAL DEFAULT       15 kill_builtin
 1870: 00000000000f5400     51 FUNC    GLOBAL DEFAULT       15 next_history
 1871: 00000000000695c0    771 FUNC    GLOBAL DEFAULT       15 dequote_string
 1872: 00000000000c0540    452 FUNC    GLOBAL DEFAULT       15 shopt_listopt
```

So there's actually a hierarchy:

1. `.symtab` which wasn't stripped, either from the loaded file or the debug file.
2. `.dynsym` from the loaded file, and a `.symtab` from the `.gnu_debugdata`.
3. `.dynsym` without the debugdata
4. Nothing at all

I probably should write this out more clearly somewhere, either in a comment, or in the documentation somewhere. Do you have any preference?

osandov replied:
Got it. A comment summarizing this would make sense in `find_module_elf_symtab()`.

brenns10 commented on libdrgn/elf_symtab.c:
```diff
@@ -148,6 +272,56 @@ find_module_elf_symtab(struct drgn_module *module)
+				if (err)
+					return err;
+
+				module->elf_symtab_pending_files = 0;
```
Ok. So in case of an error, the mask should remain set, so that errors will propagate each time a symbol lookup happens. But if we search for a symbol table and don't find it (but don't encounter any errors in the process), we should clear the mask. That makes sense, I'll change it.

brenns10 commented on libdrgn/elf_symtab.c:
```diff
@@ -148,6 +272,56 @@ find_module_elf_symtab(struct drgn_module *module)
+		if (err)
+			return err;
+
+		if (gnu_debugdata_file) {
```
Oh yeah, I agree we can get rid of the check in `load_gnu_debugdata_file()`!

brenns10 commented:
I see your point on the testing, and yeah I think code review kinda brought our attention to the trickier parts, so we wouldn't have necessarily known the best tests to write earlier. I'll work on handling those cases and any others that come to mind.

osandov commented:
Thanks for the test cases! I'll try to take a look before heading out but may not get to it until Monday.

(3) is really just a cosmetic issue with the log message, right? If so, I'm fine keeping it as is for now. It'd still be good to have a testcase with the empty (other than the undefined placeholder) `.dynsym` just to make sure we don't blow up given that it happens in practice.

Your workaround for the unused gnu_debugdata_symtab thing looks like the best we can do, fine with me.

brenns10 commented:
I think (3) wouldn't be cosmetic, if you somehow had a file with an "empty" dynamic symbol table, and another file with a non-empty one. Then drgn wouldn't be able to use the non-empty one. But I don't really know how you'd get that situation.

osandov commented:
Good point. Ok, I think it's worth handling that case, but it can be a followup fix, so don't worry about fixing it or testing it in this PR.

osandov commented:
Oh, I have a fix for that flaky timekeeping test, so ignore that.

brenns10 commented:
Added the cases for the placeholder-only symbol table, they test the behavior as it is now -- so we can adjust them once we handle those differently. Also updated the handling for removing the gnu_debugdata symbol table.

And no rush on this! I just want to unblock you as I think this may have been the last item before the small fix-up release. But Monday is great :)

osandov reviewed (approved):
LGTM, thanks! I'll merge this once tests pass.
